lib/raw_net_capture.rb

```diff
@@ -37,28 +37,64 @@ def received(data)
 
   def sent(data)
     # when there are multiple requests on the same connection, E.g. redirection, we want the last one.
-    reset if raw_received.length > 0
+    if raw_received.length > 0
+      _transactions << Transaction.new(self)
+      reset
+    end
 
     @raw_sent << data
   end
 
-  def headers
-    separator = "\r\n\r\n"
-    raw_string = @raw_received.string
-
-    if headers_end_index = raw_string.index(separator)
-      raw_string[0...(headers_end_index + separator.length)]
-    else
-      raw_string
+  def transactions
+    _transactions.tap do |transactions|
+      if raw_received.length > 0
+        transactions << Transaction.new(self)
+      end
     end
   end
 
   private
 
+  def _transactions
+    @transactions ||= []
+  end
+
   def reset
     @raw_received = StringIO.new
     @raw_sent = StringIO.new
   end
+
+  class Transaction
+    HEADER_BODY_SEPARATOR = "\r\n\r\n"
+    attr_reader :response, :request
+
+    def initialize(capture)
+      @request = Request.new(capture)
+      @response = Response.new(capture)
+    end
+
+    class ExchangePart
+      attr_reader :headers, :body
+
+      def initialize(capture)
+        raw_string = capture.send(method).string
+
+        @headers, _, @body = raw_string.partition(HEADER_BODY_SEPARATOR).map { |p| p.empty? ? nil : p  }
+      end
+    end
+
+    class Response < ExchangePart
+      def method
+        'raw_received'
```

Comment:
Didn't even know `send` worked on strings and not just symbols ... but isn't it more typical to use symbols to identify the method?